Add local datastore to persist states of LocalScope network

chenchun opened the pull request:
Signed-off-by: Chun Chen <ramichen@tencent.com>

Fixes https://github.com/docker/libnetwork/issues/461

chenchun commented:
Depending on godep updates of libkv.

mavenugo commented:
@chenchun thanks for your fast response. The CI failure is expected since the corresponding boltdb changes have not made into libkv project.

FYI, this needs a Godep update. which we can update once https://github.com/docker/libkv/pull/46 is merged

mavenugo commented on Makefile:
```diff
@@ -68,6 +68,7 @@ install-deps:
 	go get github.com/tools/godep
 	go get github.com/golang/lint/golint
 	go get github.com/mattn/goveralls
+	go get github.com/boltdb/bolt
```
This should ideally go into godep.

chenchun replied:
Yes, once bolt gets merged into libkv project, I'll update.

abronan commented:
@chenchun You added all the etcd log files, make sure to exclude the `default.etcd` folder from your commits.

abronan commented:
Needs a vendoring of `boltdb` library too with Godep.

chenchun commented:
@abronan Thanks, updated and removed them.

abronan commented:
This will replace #439

mavenugo commented on config/config.go:
```diff
@@ -5,14 +5,15 @@ import (
-	Datastore DatastoreCfg
+	Daemon                    DaemonCfg
+	Cluster                   ClusterCfg
+	Datastore, LocalDataStore DatastoreCfg
```
Shall we call these GlobalStore and LocalStore ?

chenchun replied:
Addressed this

mavenugo commented on controller.go:
```diff
@@ -109,11 +109,11 @@ type endpointTable map[types.UUID]*endpoint
+	drivers           driverTable
+	sandboxes         sandboxTable
+	cfg               *config.Config
+	store, localStore datastore.DataStore
```
globalStore and localStore ?

chenchun replied:
Addressed this

mavenugo commented on controller.go:
```diff
@@ -397,3 +402,14 @@ func (c *controller) isDriverGlobalScoped(networkType string) (bool, error) {
 	sandbox.GC()
 }
+
+func (c *controller) scope(networkType string) (driverapi.Scope, error) {
```
I dont think it is a good idea to have a scope function per networkType.
I think this should be per object. To be precise, I think we should add a `StoreScope` function in `datastore.KV` interface. Individual objects should decide if it is local or global. 
For example, we will be including sandbox as an object  in the datastore. The sandbox objects will be local even for those networks that are global in nature.

chenchun replied:
Addressed this. 
Add `DataScope() DataScope` in `datastore.KV` interface.
Add `dataScope   datastore.DataScope` in `network` struct and assign it in `controller.addNetwork`.
For KV objects like `bitseq.Handle`/`ipam.Allocator`, since they have reference to a single `datastore.DataStore` field to store states, `DataScope()` is useless for them, just return `datastore.DataScope(-1)` in `DataScope()`
For `overlay.network`, simply return `datastore.GlobalScope`.

mavenugo commented on network.go:
```diff
@@ -469,3 +469,10 @@ func (n *network) getSvcRecords() []etchosts.Record {
 	return recs
 }
+
+func (n *network) scope() (driverapi.Scope, error) {
```
Shall we change the word `Scope` to something else ?
I think the word `scope` can confuse folks for the service scope that is currently being discussed on the UX front.

chenchun replied:
If we define `DataScope()` function in `datastore.KV`, should we also move `type DataScope` from driverapi.go to datastore.go?

mavenugo replied:
Sure. it works too.

chenchun replied:
Addressed this.

mavenugo commented on store.go:
```diff
@@ -39,6 +41,37 @@ func (c *controller) initDataStore() error {
+	c.Unlock()
+	// init local datastore
+	if !c.validateLocalDatastoreConfig() {
+		return fmt.Errorf("local datastore initialization requires a valid configuration")
```
I dont think we should fail if the user hasnt configured a local store.
Since boltdb is an embedded db (without the need for any external process), we should use that as the default and have a predefined location to store the local database. The user can always override that default setting through the configuration.

chenchun replied:
Where is the perfect place, how about "/var/lib/libnetwork/boltdb.db"?

mavenugo replied:
Today we use `/var/lib/docker/network/files` in endpoint.go.
@mrjana do you think we can use the same location as the botldb default location ?

chenchun replied:
Added `defaultLocalStoreConfig` in `store.go` and used "/var/lib/docker/network/files/boltdb.db" as the default location.

mavenugo commented on store.go:
```diff
@@ -39,6 +41,37 @@ func (c *controller) initDataStore() error {
+	eps, err := c.getEndpointsFromLocalStore()
+	if err == nil {
+		c.processEndpointsUpdate(eps, nil)
+	} else {
```
Shouldnt we check for `err != datastore.ErrKeyNotFound {` ?

chenchun replied:
Addressed this.

mavenugo commented on store.go:
```diff
@@ -369,3 +377,71 @@ func (c *controller) processEndpointUpdate(ep *endpoint) bool {
+}
+
+func networkIDFromKey(key string) (string, error) {
+	if !strings.HasPrefix(key, "docker/libnetwork/endpoint/") {
```
Lets not expose the key structure outside of the object that manages the key.
For example, the key structure for endpoint must be kept as part of endpoint.go.
Hence please remove this function and reuse the `networkIDFromKey` in endpoint.go and ofcourse make it a public function.
Also, `networkIDFromKey` in endpoint.go is unused now, hence you can even change the signature to pass a key string instead of the string[] as it is currently defined.

chenchun replied:
Addressed this

mavenugo commented on store.go:
```diff
@@ -369,3 +377,71 @@ func (c *controller) processEndpointUpdate(ep *endpoint) bool {
+			delete(*prune, ep.id)
+		}
+		ep.SetIndex(epe.LastIndex)
+		if nid, err := networkIDFromKey(epe.Key); err != nil {
```
Please look at the next comment. Lets remove this function and use the `endpoint.NetworkIDFromKey` function

chenchun commented:
Also removed useless function `getNetworkFromStore` and `getEndpointFromStore` from `store.go` and added tests to test default config for local store.

chenchun commented on controller.go:
```diff
@@ -144,6 +144,11 @@ func New(cfgOptions ...config.Option) (NetworkController, error) {
+		if err := c.initLocalStore(); err != nil {
+			// Failing to initalize datastore is a bad situation to be in.
+			// But it cannot fail creating the Controller
+			log.Debugf("Failed to Initialize LocalDatastore due to %v.", err)
```
Since we added default configurations for local store, shall we return err here?

mavenugo replied:
I think so.

chenchun replied:
Addressed this.

mavenugo commented on bitseq/store.go:
```diff
@@ -70,6 +70,12 @@ func (h *Handle) Exists() bool {
+// DataScope method returns the storage scope of the datastore
+func (h *Handle) DataScope() datastore.DataScope {
+	//This method is useless currently, hence return a arbitrary value here
+	return datastore.DataScope(-1)
```
I think it should be Global as the main intention of `bitseq` is to provide a distributed guranteed unique identity out of a bitfield. ping @mrjana @aboch WDYT ?

aboch replied:
Yes, it should be global.

chenchun replied:
Addressed this.

mavenugo commented on datastore/datastore.go:
```diff
@@ -85,8 +99,17 @@ func ParseKey(key string) ([]string, error) {
+	}
+	if kv == "boltdb" {
+		dir, _ := filepath.Split(addrs)
+		if err := os.MkdirAll(dir, 0644); err != nil {
```
Shouldnt this be part of libkv NewStore operation ?

chenchun replied:
I agree, removed this. I will make a PR to libkv.

mavenugo commented on endpoint.go:
```diff
@@ -243,15 +244,15 @@ func (ep *endpoint) KeyPrefix() []string {
+	// endpoint Key structure : docker/libnetwork/endpoint/${network-id}/${endpoint-id}
+	// it's an invalid key if the key doesn't have all the 5 key elements above
+	keyElements := strings.Split(key, "/")
+	if !strings.HasPrefix(key, "docker/libnetwork/endpoint/") || len(keyElements) < 5 {
```
can you please replace `docker/libnetwork/endpoint/` with `datastore.Key(datastore.endpointKeyPrefix)` ?

chenchun replied:
Addressed this.

mavenugo commented on endpoint.go:
```diff
@@ -243,15 +244,15 @@ func (ep *endpoint) KeyPrefix() []string {
-
 	// network-id is placed at index=1. pls refer to endpoint.Key() method
-	return types.UUID(key[1]), nil
+	return strings.Split(key, "/")[3], nil
```
I have plans of changing the key length dependencies (such as 3 or 5) out of objects and make it a datastore specific issue so that we are free to change the key length in datastore without impacting the objects.

We can address this for later.  this is just FYI.

mavenugo commented on ipam/store.go:
```diff
@@ -176,3 +176,9 @@ func (a *Allocator) deleteFromStore() error {
+// DataScope method returns the storage scope of the datastore
+func (a *Allocator) DataScope() datastore.DataScope {
+	//This method is useless currently, hence return a arbitrary value here
+	return datastore.DataScope(-1)
```
ping @aboch @mrjana

aboch replied:
Should be global scope here too.

mavenugo replied:
@aboch with the new ipam extension proposal, I believe the ipam will be used both by local and globally scoped networks and objects. In that case, should we make this configurable in ipam ?

Anyways, that doesnt impact this PR  & it must be Global for now. And can be addressed along with the ipam extension proposal.

mavenugo commented on store.go:
```diff
@@ -119,13 +139,7 @@ func (c *controller) updateEndpointToStore(ep *endpoint) error {
-	c.Lock()
-	cs := c.store
-	c.Unlock()
+	cs := c.getDataStore(n.DataScope())
```
To be technically correct, it should be `ep.DataScope`

chenchun replied:
Addressed this.

mavenugo commented on store.go:
```diff
@@ -134,26 +148,11 @@ func (c *controller) updateEndpointToStore(ep *endpoint) error {
-	c.Lock()
-	cs := c.store
-	c.Unlock()
+	cs := c.getDataStore(n.DataScope())
```
same here

chenchun replied:
Addressed this.

mavenugo commented on store.go:
```diff
@@ -188,10 +187,7 @@ func (c *controller) watchNetworks() error {
-					if global {
-						tmpview[k] = v
-					}
+					tmpview[k] = v
```
This is added as part of delete processing to make sure we dont end up removing locally scoped networks incorrectly during a watch event.
With this change, we will lose that protection and the local networks will be removed when there is a watch event happens for the networks with global scope. Can you please confirm if you tried a scenario of both global and locally scoped objects ?

chenchun replied:
I changed it back. I don't know how to produce a globally and locally scoped objects given each KV object is either GlobalScope or LocalScope.

mavenugo commented on store.go:
```diff
@@ -244,28 +240,9 @@ func (n *network) watchEndpoints() error {
-					global, _ := v.network.isGlobalScoped()
-					if global {
-						tmpview[k] = v
-					}
```
Same here.

mavenugo commented:
@chenchun thanks for addressing all the comments. its shaping up nicely. I just gave a few more comments. Could you please address them ?
If you have any specific design question, we can chat in #docker-network IRC channel.

chenchun commented:
@mavenugo, I'll update the PR soon.

chenchun commented:
@mavenugo, just updated the PR. It's quite later today, if you have more comments, I'll take a look 8 hours later.

mavenugo commented on controller.go:
```diff
@@ -144,6 +144,11 @@ func New(cfgOptions ...config.Option) (NetworkController, error) {
 		}
+		if err := c.initLocalStore(); err != nil {
+			// Failing to initalize datastore is a bad situation to be in.
+			// But it cannot fail creating the Controller
```
:) pls fix the comment appropriately.

chenchun replied:
Done

mavenugo commented:
@chenchun thanks for the contribution and quick responses.
Will try out the changes end to end with docker integration once tonight to confirm the changes doesn't introduce any surprises.

mavenugo commented:
@chenchun hit a problem testing the code. With your implementation, we save the default networks such as 'none' (for null driver), 'host' (for host driver) & 'bridge' (for bridge driver). When we restart the docker daemon, it creates these network and the creation fails because these networks already exists during the daemon bootup :). If the network creation fails during the bootup, the daemon fails as well.
So, I think we need some kind of special case handling for these network.

I personally prefer to NOT save these special networks. But, our current implementation doesnt special case anything in libnetwork core. We need to get driver's help to make the decision on which network is special so that we could ignore using the datastore. Maybe we could add a new DataScope constant named `none` and also extend the data-scope capability check to the driver.
@mrjana WDYT ?

chenchun commented:
>  'none' (for null driver), 'host' (for host driver) & 'bridge' (for bridge driver)
> I personally prefer to NOT save these special networks.

I don't understand this, should we just load these networks instead of creating when restarting daemon? I assume localstore is used for preserve states for these drivers.

mavenugo commented:
@chenchun right, we could ignore the error during the creation process. But, my concern is it will alter the bootup behavior that the user come to expect. For example, if the user changes the config (say `--bip`) and restart the docker daemon, we must honor that. But, if the network create fails and we ignore the failure, then the config will be ignored as well. Hence I think the cleaner fix is to not save this config.

BTW, this behavior of changing the network configuration (during bootup) is only applicable to the special bridge driver. We cannot change the network configuration for any other user defined network.

Hope that answers.

chenchun commented:
@mavenugo how can I reach you on IRC? I have joined #docker-network channel and I'm chenchun_ on IRC.

mavenugo commented:
@chenchun am *madhu* in IRC & I responded to your messages.

mavenugo commented:
@chenchun am working on a proposal to the problem that i identified during the tests. Though it is a very special case, we are trying to address this in a generic way. Will push a PR on top of yours which you could choose to use.

chenchun commented:
@mavenugo good to know that.

mavenugo commented:
@chenchun just an update. we unearthed some other issues trying to solve the problem that I reported. currently working on resolving them. thanks for rebasing to the latest. it wont take must time from here to get it merged. fyi.

chenchun commented:
@mavenugo It's all right. I'm currently ignoring the possibilities of changing configs after restart and keep going.

mavenugo commented:
@chenchun @mrjana @aboch I have a proposed fix for the issue that I mentioned earlier in this PR. https://github.com/mavenugo/libnetwork/commit/2306f78acf400cc02c6bc4d169ff093a71ddc58c
The commit message is verbose enough to explain the problem and the proposed solution.

@chenchun could you please give this fix a try on top of your PR ?
@mrjana @aboch WDYT about the solution ?
If we agree, then we could merge this PR followed by a new PR with the above fix.

chenchun commented:
@mavenugo Cherry-picked your commit. Still not finish this PR, some other things need to be handled such as `Restore portMapping`.

mavenugo commented on controller.go:
```diff
@@ -270,6 +284,22 @@ func (c *controller) NewNetwork(networkType, name string, options ...NetworkOpti
+		n := c.getNetworkByNameFromLocalStore(name)
+		if n != nil {
+			n.generic = network.generic
+			n.endpoints = network.endpoints
```
Pls remove updating `n.endpoints` here in your latest commit. This is not being done in `setupNetwork` which is common for multiple codepath.

chenchun replied:
Fixed with latest commit.

mavenugo commented:
thanks @chenchun. 
@mrjana @aboch need your help with your review, especially on the last 2 commits 
* commit from me on the controller's `started` state : https://github.com/chenchun/libnetwork/commit/8a3fab23848cbc4867709735980803c3eb41d3a9
* commit from @chenchun in bridge driver to honor pre-populated endpoint-info

chenchun commented:
@mavenugo there are some problems with tests now. I'll address them.

mrjana commented on store.go:
```diff
@@ -6,62 +6,133 @@ import (
+		if err == nil {
+			c.processNetworkUpdate(nws, nil)
+		} else if err != datastore.ErrKeyNotFound {
+			log.Warnf("failed to read networks from globalstore during init : %v", err)
```
Why are we not iterating endpoints from globalstore while we do it in localstore? In general for configuration both the stores should be treated exactly identically. In fact looking at this whole diff I am not a fan of having boiler plate code which does the same thing separately for globalstore and localstore. Once we decide and agree that the objects from global and local stores needs to be processed exactly the same way then we need to have common functions which operate on a list of stores and that's it. One of the property of the store could be locality which could be used for object placement purposes i.e another common function which WriteToSore but based on a KVObject Interface method called `Scope` will choose the right store to write the object into.

chenchun replied:
Added `updateToStore`/`deleteFromStore` and removed `updateNetworkToStore `/`updateEndpointToStore `/`deleteNetworkFromStore `/`deleteEndpointFromStore `

mrjana commented on controller.go:
```diff
@@ -270,6 +285,21 @@ func (c *controller) NewNetwork(networkType, name string, options ...NetworkOpti
 	network.processOptions(options...)
 
+	if !c.started {
+		// Global stores are watched only when the controller enters the 'started' state.
```
Does this comment assume too much about how the caller is using libnetwork? Why can't globalNetworks before Start? Even assuming existing docker side implementation what happens when the default network is configured to be a global network? Also I don't see the equivalent reconciliation for endpoints

chenchun replied:
I'm not sure how to update this comment. @mavenugo, can you take a look?
> Also I don't see the equivalent reconciliation for endpoints

After restart docker daemon will only recreate networks, so I think reconciliation for endpoints is not needed?

mavenugo replied:
@mrjana valid points. but this was added with some inbuilt assumptions and existing behavior.

1. we dont create default-network (other than docker0) during system init. Default-networks are created lazily during container activation.
2. endpoints are not created during the init phase.
3. Global networks will not be created in this phase (watch isnt enabled yet).

aboch commented on bitseq/store.go:
```diff
@@ -70,6 +70,11 @@ func (h *Handle) Exists() bool {
 
+// DataScope method returns the storage scope of the datastore
+func (h *Handle) DataScope() datastore.DataScope {
+	return datastore.GlobalScope
```
When asked about this I said bitseq ds should be global scope, but actually it makes more sense this property is configurable by who instantiates bitseq. We can have an local or global scoped id manager.

As @mavenugo said this is not a concern of this PR. It's an incremental change after this PR is merged.

aboch commented on controller.go:
```diff
@@ -157,11 +160,23 @@ func New(cfgOptions ...config.Option) (NetworkController, error) {
 
+// Start method starts the controller
+func (c *controller) Start() error {
+	if c.started {
```
Can there be any concurrent call to the method `Start()`, also exercisable via an ad-hoc test ?
Just wondering whether locking would be justifiable here.

chenchun replied:
Fixed.

aboch commented on store.go:
```diff
@@ -6,62 +6,133 @@ import (
-	n.Unlock()
+	// Check if a network already exists with the specified network name
+	c.Lock()
+	for _, cn := range c.networks {
```
Please use `(c *controller) NetworkByName() (network, error)` here.

chenchun replied:
Fixed.

chenchun commented:
In order to make rebase & review easier, I only rebased the latest 2 commits. If it's necessary I'll rebase all the commits to make them clean after getting ready to merge.
@aboch @mrjana @mavenugo 
Addressed your comments and rewrited tests. PTAL .

chenchun commented on controller.go:
```diff
@@ -301,17 +333,25 @@ func (c *controller) addNetwork(n *network) error {
 	}
-	if err := n.watchEndpoints(); err != nil {
-		return err
+	if n.isGlobalScoped() {
```
Added a check here. I think it's necessary because both local/global network addition will invoke this function.

mavenugo commented on store.go:
```diff
@@ -6,96 +6,139 @@ import (
-	if err != nil || !global {
+func (c *controller) newNetworkFromStore(n *network) error {
+	// Check if a network already exists with the specified network name
+	c.Lock()
```
Please remove the c.Lock() and Unlock() in this method. It causes deadlock :(

chenchun replied:
Fixed. And tested against docker, seems ok now.

mavenugo commented on drivers/bridge/bridge.go:
```diff
@@ -1451,3 +1479,60 @@ func electMacAddress(epConfig *endpointConfiguration, ip net.IP) net.HardwareAdd
+	// Restore container interface name
+	sbox, err := linkByMacAddress(iface.MacAddress().String())
+	if err != nil {
+		return err
```
Testing the final changes by building a docker binary. It failed with the following error when I tried to restore a bunch of saved endpoints.

```
ERRO[0000] failed to create endpoint s3 on network network1: Link with hardware addr 02:42:ac:19:00:01 not found
ERRO[0000] failed to create endpoint s4 on network network1: Link with hardware addr 02:42:ac:19:00:02 not found
ERRO[0000] failed to create endpoint c1 on network bridge: Link with hardware addr 02:42:ac:11:00:02 not found
ERRO[0000] failed to create endpoint s5 on network network2: Link with hardware addr 02:42:ac:1a:00:01 not found
```

mavenugo replied:
@chenchun am wondering how you are testing these changes ? am facing all sorts of issues when I integrated it with docker daemon. am a bit concerned now that this PR is trying to do too many things.
Also, it would be good if you can explain the use-case that you are trying to address with this one. I remember you mentioning about the hot-restart of the docker daemon. But AFAIK it is not targetted for 1.9 release. So, it would be good if you can clarify the testing methodology and the use-case clearly asap.

chenchun replied:
I tested it using unit tests in the code mainly. I have tested aginst docker before I introduced restoring portmapping/macadress. I apologize for the errors you encounted. I should be more careful. For the user case, I have nothing else except for hot upgrade. And it seems you have something depending on localstore, so I suggest to split the last commit.

mavenugo replied:
@chenchun thanks for the confirmation. just for my information, when you say `hot upgrade`, you mean restarting daemon while containers are still running. Correct ?
am not sure if docker 1.9 is going to support this feature. Is it something that you are playing with in private ?

chenchun replied:
yes, I'm trying to implement that at my local branch https://github.com/chenchun/docker/tree/upgrades?files=1. We want to implement that ASAP to enhance the stablity of our resources schedule system.

mavenugo replied:
thanks. can you please fix these issues and make sure the basic use-case works first ?
our main motivation is to provide save / restore of network & endpoint states for local scoped objects so that we get all the info as before restarting the docker daemon. This is not a hot-swap case. But, it it breaks the basic use-case, then we will have hard time getting this in for 1.9.

chenchun replied:
The branch is out of date with the master branch of docker. I have not rebased because I wait to integrate localstore feather from libnetwork. I want to implement hot upgrade because we encountered many bugs of docker and we have more requirements as we are running more kinds of apps from the other teams in my company that we can't afford to stop daemon & kill all containers in order to upgrade docker binary.

mavenugo replied:
@chenchun that is okay. can you please make sure the basic use-case works fine in 1.9 without any of your hot-upgrade changes ? That way you don't have to worry about master branch in docker out-of-date for your needs. I do understand that if you want to also include the hot-restart testing in the same branch, then it might be difficult for you.
Please note that we are prioritizing the PR  for the reason that it is having a positive impact on the release. If it impacts the base docker release code, we wont be able to accept it till it is resolved properly. Hope you understand that.

chenchun commented:
I agree it tries to do too many things. I think we can split the last commit and focus on introducing localstore. But these should be a way to handle https://github.com/docker/libnetwork/blob/master/drivers/bridge/bridge.go#L887

mavenugo commented:
thanks @chenchun. I noticed that you have reverted the bridge driver changes that I reported. Is that expected ? Without the changes to the bridge driver, wouldnt restore fail as I indicated earlier ?
I will not be able to give this patch a try tonight. But it will be good if you can give me some description of the latest changes that you have done to enable with the code-review.

mavenugo commented:
@chenchun based on our discussion, i agree that the bridge driver changes aren't required & also the daemon restart is going to cleanup the endpoint anyways.
I discussed this issue with the other maintainers and we feel that it is infact better to not even save the endpoint for local scope (we still need it in the globalstore).
Can you please make that change ? And I think that will solve all the issues and makes it is ready for merge.

chenchun commented:
> also the daemon restart is going to cleanup the endpoint anyways

Correct me if I'm wrong. I think daemon restarts won't delete endpoints currently, because daemon only keeps SandboxID/EndpointID now and when daemon is killed manually or machine crashes, daemon won't be able to delete veth devices. So I think it's better to keep endpoints and deletes them upon restart.

mavenugo commented:
@chenchun you are right. But, we have other ideas to clean up the stale endpoints. Another reason to keep the endpoints deletion away from this PR is to reduce the complexity and get this in ASAP.
So, it would be good if we get this PR in with network save/restore right away and work on the endpoint cleanup on bootup in docker. Also, we have multiple PRs pending on the libkv upgrade. Shall we get this PR to get Network save and restore completed followed by another for proper endpoint handling ?
am in IRC and we can work together on this.

mavenugo commented on drivers/remote/driver.go:
```diff
@@ -52,9 +53,9 @@ func (d *driver) getCapabilities() (*driverapi.Capability, error) {
+		c.DataScope = datastore.GlobalScope
 	case "local":
-		c.Scope = driverapi.LocalScope
+		c.DataScope = datastore.LocalScope
```
Thats a very quick rebase :-).

chenchun replied:
Unfortunately forget to refactor tests during rebase.

mavenugo commented:
Thanks @chenchun for the contribution and addressing all the comments patiently.
I tested e2e with both local and global store.

LGTM.

ping @mrjana @aboch

aboch commented:
LGTM

mavenugo commented:
@chenchun as we discussed, we were trying to simplify this logic by removing my commit from this PR. With the introduction of https://github.com/docker/libnetwork/pull/547, we can actually do that ... :fireworks: 

Shall I request you to submit the PR again (with a rebase) and remove my commit (https://github.com/chenchun/libnetwork/commit/652d319c33bef98e58f1223b6f47d25a609f1291) from this PR ? BTW I attempted this change in : https://github.com/mavenugo/libnetwork/commits/defnet_save. Please use it if you find it useful. (You can ignore my last commit that introduces `persist` option),

This ofcourse needs changes in the docker daemon to call controller.New with driver init params. Which we will take care during vendoring in.

chenchun commented:
OK, will update the PR shortly.

mavenugo commented:
Thanks @chenchun pls let me know if you need any help.

chenchun commented:
@mavenugo, some problems with unit tests now. Since we changed many unit tests to pass configs to `controller.New` function, some tests will create localstore now. And cause they are run in a single process, they will hang on creating the same boltdb file lock.

mavenugo commented:
@chenchun. Can't we use config Params to point to different Boltdb file for each of the different controllers to make use of ?

chenchun commented:
Yes, doing it right now.

chenchun commented:
@mavenugo updated now. Will test it against docker tomorrow.

chenchun commented on network.go:
```diff
@@ -233,13 +240,22 @@ func (n *network) Delete() error {
 		return err
 	}
 
+	defer func() {
```
Add a defer here to restore network.

chenchun commented on endpoint.go:
```diff
@@ -394,27 +399,31 @@ func (ep *endpoint) Delete() error {
 	}
 	defer func() {
 		if err != nil {
-			ep.SetIndex(0)
```
Changed `ep.SetIndex(0)` to `ep.dbExists = false` as like in `network.Delete`.

mavenugo commented:
@chenchun thanks for your patience and handling all the comments.
We still need a couple of fixes to make it work properly e2e with Docker Daemon.
(such as https://github.com/mavenugo/libnetwork/commit/76c9c1369633ecfc13d91174698d6335004dc36f). But we can handle it via subsequent PRs.

Will merge this PR.

chenchun commented:
:+1: , Happy to see it finally gets merged. I apologize if I slow down your work during the process of this PR. Thanks for your patient review and help too.
